tests/integration/v3_watch_test.go

```diff
@@ -1467,3 +1468,48 @@ func TestV3WatchProgressWaitsForSync(t *testing.T) {
 		t.Fatal("Wrong revision in progress notification!")
 	}
 }
+
+func TestV3WatchProgressWaitsForSyncNoEvents(t *testing.T) {
+	if integration.ThroughProxy {
+		t.Skip("grpc proxy currently does not support requesting progress notifications")
+	}
+	integration.BeforeTest(t)
+
+	clus := integration.NewCluster(t, &integration.ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	client := clus.RandClient()
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	resp, err := client.Put(ctx, "bar", "1")
+	require.NoError(t, err)
+
+	wch := client.Watch(ctx, "foo", clientv3.WithRev(resp.Header.Revision))
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	err = client.RequestProgress(ctx)
```

Comment:
If I understand it correctly, the `client.RequestProgress` won't get any progress notification response before etcd moves the watcher from unsynced list to synced list (see `(s *watchableStore) syncWatchers`), but it will set the `deferredProgress` to `true`.  But there isn't any data/events to send for the key "foo", so the `deferredProgress` will never to set to `false`, accordingly all the following `client.RequestProgress` in the loop won't get any progress notification without the fix in this PR.

Please add a comment in the test case.

Reply:
Added a comment why we send a progress notification here.